packages/webapp/src/hooks/useOfflineCheck.ts

```diff
@@ -0,0 +1,73 @@
+import { useState, useEffect } from "react";
+import { toast } from "sonner";
+
+async function performNetworkCheck() {
+  const endpoint = "https://www.google.com";
+  try {
+    const response = await fetch(endpoint, {
+      method: "HEAD",
+      cache: "no-cache",
+      mode: "no-cors",
+    });
+    return response.ok;
+  } catch (error) {
+    throw error; // Rethrow for retries
+  }
+}
+
+function useOfflineCheck(options: any = {}) {
+  const {
+    // Optional configuration properties:
+    threshold = 1000, // Minimum interval between offline checks (ms)
+    maxRetries = 3, // Maximum number of retries before assuming offline
+    retryDelay = 1500, // Delay between retries (ms)
+  } = options;
+
+  const [isOffline, setIsOffline] = useState(false);
+  const [retries, setRetries] = useState(0);
+
+  useEffect(() => {
+    const handleOnline = () => {
+      setIsOffline(false);
+      setRetries(0); // Reset retries on reconnection
+    };
+
+    const handleOffline = () => {
+      if (retries < maxRetries) {
+        setTimeout(() => {
+          performNetworkCheck()
+            .then(() => handleOnline()) // online
+            .catch(() => {
+              setRetries(retries + 1); // Retry if error occurs
+            });
+        }, retryDelay);
+      } else {
+        setIsOffline(true);
+        toast.error("App is offline. Please check your internet connection.");
+      }
+    };
+
+    window.addEventListener("online", handleOnline);
+    window.addEventListener("offline", handleOffline);
+
+    // Initial network check:
+    performNetworkCheck()
+      .then(() => {
+        setIsOffline(false);
+        setRetries(0);
+      })
+      .catch(() => {
+        setIsOffline(true);
+        toast.error("App is offline. Please check your internet connection.");
+      });
```

Comment:
The toast is triggered many times, which we should fix.

I'm wondering if this isn't the problem, as this is rerun every time `useOfflineCheck` is rerun, which is every time that the page is rendered.

Another possible source of the issue is that the "offline" event is fired multiple times? Is that the case?

Reply:
i think i figured why the toast is rendering multiple times, it is being called both in the `performNetworkCheck()` function and the handleOffline.

Reply:
Did you try the codebase and shut down your wifi to see if this works correctly?